synapse/push/httppusher.py

```diff
@@ -12,10 +14,17 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
+import hashlib
+import hmac
+import json
 import logging
 import urllib.parse
 from typing import TYPE_CHECKING, Any, Dict, Iterable, Optional, Union
 
+import unpaddedbase64
+from Crypto.Cipher import AES
+from Crypto.Util.Padding import pad
+from donna25519 import PrivateKey, PublicKey  # type: ignore
```

Comment:
Woops, forgot about the type-ignore. The reason here being is that donna25519 doesn't have typings and at the time of writing the code soru didn't have time to create them. If wanted, she can do that!

Soru did originally plan to first use pynacl, it, however, doese not work: pynacl's Box and SealedBox seems to do some more things under the hood (key expansion? no idea...) so that soru could not get it to work with existing / other implementations of the algorithm described in the MSC to work (namely, libolms PkEncryption / PkDecryption), which is why soru begrudgingly pulled in donna25519 instead.

Reply:
Thanks, that makes sense.

(It's fine that there aren't annotations for donna25519; I'm just surprised that the type-ignore is needed in that case. I think we typically add an `ignore_missing_imports = True` line at the bottom of `mypy.ini` which eliminates the need for this.)

Reply:
Another thought: donna25519 is licensed with the "Unlicense". Are we okay to add this as a dependency @callahad? My broad understanding is that any FOSS licence that isn't the GPL is compatible with Synapse's Apache 2 licence.

Reply:
Fixed the import typeignore in mypy.ini, leaving this thread open for the license question

Reply:
Unlicense is an OSI Approved free software license, and it's non-copyleft / non-viral, so that works for me... but I am a little uncomfortable with that dependency given that the upstream author of curve25519-donna [discourages its use](https://github.com/agl/curve25519-donna#curve25519-donna):

> Note: this code is from 2008. Since that time, many more, great implementations of curve25519 have been written, including several amd64 assembly versions by djb. You are probably better served now by NaCl or libsodium.

Reply:
> Note: this code is from 2008. Since that time, many more, great implementations of curve25519 have been written, including several amd64 assembly versions by djb. You are probably better served now by NaCl or libsodium.

PyNaCl does not work, as outlined above, sadly. Also, while that code may be from 2008, the underlying crypto native doesn't suddenly change, if that makes sense?

Reply:
Unfortunately, I consider this a blocker: we need to find a way to use nacl or libsodium for curve25519 here.

Additionally, we need to run well on many operating systems and architectures, but donna25519 is unmaintained and only provides wheels for CPython 3.4/3.5 on Windows: https://pypi.org/project/donna25519/#files :(

Reply:
Is there any other maintained curve25519 implementation we can get it? Does OpenSSL / `cryptography` expose anything usable?

Reply:
> while that code may be from 2008, the underlying crypto native doesn't suddenly change, if that makes sense?

It does; I'm sure everything is fine from a cryptographic standpoint.

I'm more concerned about the ongoing maintenance and performance costs of depending on an unmaintained Python library which is itself a wrapper around an unmaintained C library. :)

Reply:
`cryptography` has support for this in its 'hazardous materials' package...
```
from cryptography.hazmat.primitives.asymmetric.x25519 import X25519PrivateKey, X25519PublicKey

...
    print('==== cryptography hazmat ====')
    haz_key_1 = X25519PrivateKey.from_private_bytes(key_1)
    haz_key_1_pub = X25519PublicKey.from_public_bytes(key_1_pub)
    haz_key_2 = X25519PrivateKey.from_private_bytes(key_2)
    haz_key_2_pub = X25519PublicKey.from_public_bytes(key_2_pub)
    print(unpaddedbase64.encode_base64(haz_key_1.exchange(haz_key_2_pub)))
    print(unpaddedbase64.encode_base64(haz_key_2.exchange(haz_key_1_pub)))
```


```
==== donna25519 exchanges ====
aCdaJq7TQ6ECo/1OrZ7u/RN5yyElbtw1Hphi0gG7Lls
aCdaJq7TQ6ECo/1OrZ7u/RN5yyElbtw1Hphi0gG7Lls
==== pynacl exchanges ====
C/VgMbdGNcdn8F9YDQTVQ5rWwslZAFPCexgU1ZRwzR0
C/VgMbdGNcdn8F9YDQTVQ5rWwslZAFPCexgU1ZRwzR0
==== cryptography hazmat ====
aCdaJq7TQ6ECo/1OrZ7u/RN5yyElbtw1Hphi0gG7Lls
aCdaJq7TQ6ECo/1OrZ7u/RN5yyElbtw1Hphi0gG7Lls
```

(I couldn't quickly see why NaCl disagrees, but I didn't dig into the C source.)

Reply:
Soru had actually asked on pynacl and had asked them about this, and they also recommended using `cryptography` https://github.com/pyca/pynacl/issues/720, will look into updating this PR soon

Reply:
soru tested and verified that the cryptography module is working, so she replaced donna25519 with cryptography now